### PR TITLE
Cache MethodToCallFromOutside and create blueprints in parallel

### PR DESCRIPTION
**Pull request by srmurguia**
The function getMethodToCallForDependency is called on every dependency.
This function creates a new ModuleBluePrint to extract
methodToCallFromOutside. In a project with N modules there can be upto
N*(N-1)/2 dependencies, meaning that the same number of ModuleBluePrint will
be generated and deleted.

This change adds a cache to store the values of methodToCallFromOutside,
needing now only N ModuleBluePrint.

Also, now the constructor of ProjectBlueprint create moduleBlueprints in
parallel.

A project with 1000 modules went from taking around 100 seconds to just
18 (5x imrpovement in time).

### CODE REVIEW

**Comment by NikitaKozlov** [comment]
I'll take a closer look over the weekend, but I think that if we want to measure more stuff we should something less explicit. Something based on AspectJ, for example. I wrote a library once for that, we don't have to use it, but it was very nice that AspectJ helped to hide these measurements. Probably we don't need to do it for this PR, though

**Comment by NikitaKozlov** [comment]
Please bring master to your branch, we have CI now :)

File: src/main/kotlin/com/google/androidstudiopoet/ModuleBlueprintFactory.kt [hunk 1]
```diff
@@ -24,7 +28,15 @@ object ModuleBlueprintFactory {
             dependency, return methodToCallFromOutside and forget about this module blueprint.
             WARNING: creation of ModuleBlueprint could be expensive
          */
-        return ModuleBlueprint(index, getModuleNameByIndex(index), projectRoot, listOf(), listOf(), config).methodToCallFromOutside
+        synchronized(methodLock[index]) {
```
Comment (NikitaKozlov) [inline src/main/kotlin/com/google/androidstudiopoet/ModuleBlueprintFactory.kt]: Can you please extract this whole block into a separate method? You can call it something like `getMethodToCallForModule`, then we'll be able to optimize it even more without touching the main method.

Comment (srmurguia) [reply src/main/kotlin/com/google/androidstudiopoet/ModuleBlueprintFactory.kt]: Thanks for this suggestion, it makes a lot of sense to have it in a function. Hopefully this function will be faster than creating and throwing a ModuleBlueprint once we implement it :p

Comment (NikitaKozlov) [reply src/main/kotlin/com/google/androidstudiopoet/ModuleBlueprintFactory.kt]: In the long run, I hope we can build modules in the order that doesn't require to have a cache at all unless there is a circular dependency. :)

**Pull request merged**